File: LeetCode_Solutions/LeetCode_Solutions/HappyNumber_202.m
//
//  HappyNumber_202.m
//  LeetCode_Solutions
//
//  Created by 来自遥远星系的核心巡洋舰 on 16/11/7.
//  Copyright © 2016年 KnightJoker. All rights reserved.
//


#import "HappyNumber_202.h"

@implementation HappyNumber_202

- (BOOL)isHappy:(NSInteger)n {

    NSInteger count = 0;
    
    if (n > 10 || n == 10) {
        NSInteger temp = n % 10;
        n = n / 10;
        
        if (n > 0 && n < 10) {
            count = temp * temp + n * n + count;
            
            if (count == 1) {
                return YES;
            } else {
                [self isHappy:count];
            }
            
        } else {
            [self isHappy:n];
        }
     
    }
    return NO;
    
}

@end

File: LeetCode_Swift/LeetCode_Swift/HappyNumber_202.swift
//
//  HappyNumber_202.swift
//  LeetCode_Swift
//
//  Created by Huni on 8/19/17.
//  Copyright © 2017 KnighhtJoker. All rights reserved.
//

import Foundation

public class HappyNumber_202 {

    func isHappy(_ n: Int) -> Bool {
        
        var set = Set<Int>()
        var num = n
        
        while num != 1 {
            guard !set.contains(num) else {
                return false
            }
            set.insert(num)
            
            var temp = 0
            while num != 0 {
                temp += (num % 10) * (num % 10)
                num = num / 10
            }
            num = temp
        }
        
        return true
    }
}

